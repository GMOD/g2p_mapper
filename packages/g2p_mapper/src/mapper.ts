// see similar function in msaview plugin
export function genomeToTranscriptSeqMapping(feature: Feat) {
  const strand = parseStrand(feature.strand)
  const refName = feature.seq_id
  const cds =
    feature.child_features
      // @ts-expect-error
      ?.map(f => f[0])
      ?.filter(f => f.type === 'CDS')
      .sort((a, b) => strand * (a.start - b.start)) || []
  const g2p = {} as Record<number, number | undefined>
  const p2g = {} as Record<number, number | undefined>

  let proteinCounter = 0
  if (strand !== -1) {
    for (const f of cds) {
      for (let genomePos = f.start; genomePos < f.end; genomePos++) {
        const proteinPos = Math.floor(proteinCounter++ / 3)
        g2p[genomePos] = proteinPos
        if (!p2g[proteinPos]) {
          p2g[proteinPos] = genomePos
        }
      }
    }
  } else {
    for (const f of cds) {
      for (let genomePos = f.end; genomePos > f.start; genomePos--) {
        const proteinPos = Math.floor(proteinCounter++ / 3)
        g2p[genomePos] = proteinPos
        if (!p2g[proteinPos]) {
          p2g[proteinPos] = genomePos
        }
      }
    }
  }

  return { g2p, p2g, refName, strand }
}
