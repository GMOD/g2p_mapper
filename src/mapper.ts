export interface Feat {
  strand: number
  refName: string
  type: string
  phase: number
  id: string
  start: number
  end: number
  subfeatures?: Feat[]
}

export function genomeToTranscriptSeqMapping(feature: Feat) {
  const strand = feature.strand
  const refName = feature.refName

  // Filter CDS features and deduplicate based on start/end positions
  // (GFF3 files can contain duplicate CDS features)
  const cdsFeatures = feature.subfeatures?.filter(f => f.type === 'CDS') ?? []
  const uniqueCdsMap = new Map<string, Feat>()
  for (const f of cdsFeatures) {
    const key = `${f.start}-${f.end}`
    if (!uniqueCdsMap.has(key)) {
      uniqueCdsMap.set(key, f)
    }
  }
  const cds = Array.from(uniqueCdsMap.values()).sort(
    (a, b) => strand * (a.start - b.start),
  )

  const g2p = {} as Record<number, number>
  const p2g = {} as Record<number, number>

  let proteinCounter = 0
  if (strand !== -1) {
    for (const f of cds) {
      for (let genomePos = f.start; genomePos < f.end; genomePos++) {
        const proteinPos = Math.floor(proteinCounter++ / 3)
        g2p[genomePos] = proteinPos
        if (!p2g[proteinPos]) {
          p2g[proteinPos] = genomePos
        }
      }
    }
  } else {
    // For reverse strand, iterate from end-1 down to start (inclusive)
    // to properly handle 0-based half-open intervals [start, end)
    for (const f of cds) {
      for (let genomePos = f.end; genomePos >= f.start; genomePos--) {
        const proteinPos = Math.floor(proteinCounter++ / 3)
        g2p[genomePos] = proteinPos
        if (!p2g[proteinPos]) {
          p2g[proteinPos] = genomePos
        }
      }
    }
  }

  return { g2p, p2g, refName, strand }
}
