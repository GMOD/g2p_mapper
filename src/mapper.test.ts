import { test, expect } from 'vitest'
import { genomeToTranscriptSeqMapping } from './mapper'

test('basic', () => {
  const ret = genomeToTranscriptSeqMapping({
    type: 'mRNA',
    id: 'hello',
    start: 100,
    end: 200,
    strand: 1,
    phase: 0,
    refName: 'chr1',
    subfeatures: [
      {
        type: 'CDS',
        start: 100,
        end: 200,
        id: 'hello1',
        phase: 0,
        strand: 1,
        refName: 'chr1',
      },
    ],
  })
  expect(ret).toMatchSnapshot()
})

test('reverse strand gene with duplicate CDS features', () => {
  // This test case has duplicate CDS features (which can occur in GFF3 files)
  // and tests that:
  // 1. Duplicates are handled correctly (not processed twice)
  // 2. Reverse strand mapping starts at protein position 0 (not 10)
  // 3. Boundary positions are included correctly
  const ret = genomeToTranscriptSeqMapping({
    type: 'mRNA',
    id: 'rna-XM_034563908.1',
    start: 438289,
    end: 440094,
    strand: -1,
    phase: 0,
    refName: 'NC_036159.2',
    subfeatures: [
      // First set of CDS features
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 438289,
        end: 438320,
        strand: -1,
        phase: 1,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 438638,
        end: 438672,
        strand: -1,
        phase: 2,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 439388,
        end: 439439,
        strand: -1,
        phase: 2,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 439581,
        end: 439691,
        strand: -1,
        phase: 1,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 439839,
        end: 439935,
        strand: -1,
        phase: 1,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 440062,
        end: 440094,
        strand: -1,
        phase: 0,
        refName: 'NC_036159.2',
      },
      // Exon (should be ignored)
      {
        type: 'exon',
        id: 'exon-XM_034563908.1-6',
        start: 438289,
        end: 438320,
        strand: -1,
        phase: 0,
        refName: 'NC_036159.2',
      },
      // Duplicate CDS features (should be deduplicated)
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 438289,
        end: 438320,
        strand: -1,
        phase: 1,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 438638,
        end: 438672,
        strand: -1,
        phase: 2,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 439388,
        end: 439439,
        strand: -1,
        phase: 2,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 439581,
        end: 439691,
        strand: -1,
        phase: 1,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 439839,
        end: 439935,
        strand: -1,
        phase: 1,
        refName: 'NC_036159.2',
      },
      {
        type: 'CDS',
        id: 'cds-XP_034419714.1',
        start: 440062,
        end: 440094,
        strand: -1,
        phase: 0,
        refName: 'NC_036159.2',
      },
    ],
  })

  // For reverse strand, the first CDS in transcription order is at the highest genomic position
  // The last CDS (440062-440094) should map to the start of the protein (position 0)
  // Position 440093 (end-1) should map to protein position 0 (first codon starts here)
  expect(ret.g2p[440093]).toBe(0)

  // The first position (438289) should also be included
  expect(ret.g2p[438289]).toBeDefined()

  // Check that we're not seeing protein position 10 at the end position
  // (which would indicate duplicates were processed)
  expect(ret.g2p[440093]).not.toBe(10)

  expect(ret).toMatchSnapshot()
})
