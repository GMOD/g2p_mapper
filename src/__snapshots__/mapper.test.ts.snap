// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
{
  "g2p": {
    "100": 0,
    "101": 0,
    "102": 0,
    "103": 1,
    "104": 1,
    "105": 1,
    "106": 2,
    "107": 2,
    "108": 2,
    "109": 3,
    "110": 3,
    "111": 3,
    "112": 4,
    "113": 4,
    "114": 4,
    "115": 5,
    "116": 5,
    "117": 5,
    "118": 6,
    "119": 6,
    "120": 6,
    "121": 7,
    "122": 7,
    "123": 7,
    "124": 8,
    "125": 8,
    "126": 8,
    "127": 9,
    "128": 9,
    "129": 9,
    "130": 10,
    "131": 10,
    "132": 10,
    "133": 11,
    "134": 11,
    "135": 11,
    "136": 12,
    "137": 12,
    "138": 12,
    "139": 13,
    "140": 13,
    "141": 13,
    "142": 14,
    "143": 14,
    "144": 14,
    "145": 15,
    "146": 15,
    "147": 15,
    "148": 16,
    "149": 16,
    "150": 16,
    "151": 17,
    "152": 17,
    "153": 17,
    "154": 18,
    "155": 18,
    "156": 18,
    "157": 19,
    "158": 19,
    "159": 19,
    "160": 20,
    "161": 20,
    "162": 20,
    "163": 21,
    "164": 21,
    "165": 21,
    "166": 22,
    "167": 22,
    "168": 22,
    "169": 23,
    "170": 23,
    "171": 23,
    "172": 24,
    "173": 24,
    "174": 24,
    "175": 25,
    "176": 25,
    "177": 25,
    "178": 26,
    "179": 26,
    "180": 26,
    "181": 27,
    "182": 27,
    "183": 27,
    "184": 28,
    "185": 28,
    "186": 28,
    "187": 29,
    "188": 29,
    "189": 29,
    "190": 30,
    "191": 30,
    "192": 30,
    "193": 31,
    "194": 31,
    "195": 31,
    "196": 32,
    "197": 32,
    "198": 32,
    "199": 33,
  },
  "p2g": {
    "0": 100,
    "1": 103,
    "10": 130,
    "11": 133,
    "12": 136,
    "13": 139,
    "14": 142,
    "15": 145,
    "16": 148,
    "17": 151,
    "18": 154,
    "19": 157,
    "2": 106,
    "20": 160,
    "21": 163,
    "22": 166,
    "23": 169,
    "24": 172,
    "25": 175,
    "26": 178,
    "27": 181,
    "28": 184,
    "29": 187,
    "3": 109,
    "30": 190,
    "31": 193,
    "32": 196,
    "33": 199,
    "4": 112,
    "5": 115,
    "6": 118,
    "7": 121,
    "8": 124,
    "9": 127,
  },
  "refName": "chr1",
  "strand": 1,
}
`;

exports[`reverse strand gene with duplicate CDS features 1`] = `
{
  "g2p": {
    "438290": 117,
    "438291": 117,
    "438292": 117,
    "438293": 116,
    "438294": 116,
    "438295": 116,
    "438296": 115,
    "438297": 115,
    "438298": 115,
    "438299": 114,
    "438300": 114,
    "438301": 114,
    "438302": 113,
    "438303": 113,
    "438304": 113,
    "438305": 112,
    "438306": 112,
    "438307": 112,
    "438308": 111,
    "438309": 111,
    "438310": 111,
    "438311": 110,
    "438312": 110,
    "438313": 110,
    "438314": 109,
    "438315": 109,
    "438316": 109,
    "438317": 108,
    "438318": 108,
    "438319": 108,
    "438320": 107,
    "438639": 107,
    "438640": 107,
    "438641": 106,
    "438642": 106,
    "438643": 106,
    "438644": 105,
    "438645": 105,
    "438646": 105,
    "438647": 104,
    "438648": 104,
    "438649": 104,
    "438650": 103,
    "438651": 103,
    "438652": 103,
    "438653": 102,
    "438654": 102,
    "438655": 102,
    "438656": 101,
    "438657": 101,
    "438658": 101,
    "438659": 100,
    "438660": 100,
    "438661": 100,
    "438662": 99,
    "438663": 99,
    "438664": 99,
    "438665": 98,
    "438666": 98,
    "438667": 98,
    "438668": 97,
    "438669": 97,
    "438670": 97,
    "438671": 96,
    "438672": 96,
    "439389": 96,
    "439390": 95,
    "439391": 95,
    "439392": 95,
    "439393": 94,
    "439394": 94,
    "439395": 94,
    "439396": 93,
    "439397": 93,
    "439398": 93,
    "439399": 92,
    "439400": 92,
    "439401": 92,
    "439402": 91,
    "439403": 91,
    "439404": 91,
    "439405": 90,
    "439406": 90,
    "439407": 90,
    "439408": 89,
    "439409": 89,
    "439410": 89,
    "439411": 88,
    "439412": 88,
    "439413": 88,
    "439414": 87,
    "439415": 87,
    "439416": 87,
    "439417": 86,
    "439418": 86,
    "439419": 86,
    "439420": 85,
    "439421": 85,
    "439422": 85,
    "439423": 84,
    "439424": 84,
    "439425": 84,
    "439426": 83,
    "439427": 83,
    "439428": 83,
    "439429": 82,
    "439430": 82,
    "439431": 82,
    "439432": 81,
    "439433": 81,
    "439434": 81,
    "439435": 80,
    "439436": 80,
    "439437": 80,
    "439438": 79,
    "439439": 79,
    "439582": 79,
    "439583": 78,
    "439584": 78,
    "439585": 78,
    "439586": 77,
    "439587": 77,
    "439588": 77,
    "439589": 76,
    "439590": 76,
    "439591": 76,
    "439592": 75,
    "439593": 75,
    "439594": 75,
    "439595": 74,
    "439596": 74,
    "439597": 74,
    "439598": 73,
    "439599": 73,
    "439600": 73,
    "439601": 72,
    "439602": 72,
    "439603": 72,
    "439604": 71,
    "439605": 71,
    "439606": 71,
    "439607": 70,
    "439608": 70,
    "439609": 70,
    "439610": 69,
    "439611": 69,
    "439612": 69,
    "439613": 68,
    "439614": 68,
    "439615": 68,
    "439616": 67,
    "439617": 67,
    "439618": 67,
    "439619": 66,
    "439620": 66,
    "439621": 66,
    "439622": 65,
    "439623": 65,
    "439624": 65,
    "439625": 64,
    "439626": 64,
    "439627": 64,
    "439628": 63,
    "439629": 63,
    "439630": 63,
    "439631": 62,
    "439632": 62,
    "439633": 62,
    "439634": 61,
    "439635": 61,
    "439636": 61,
    "439637": 60,
    "439638": 60,
    "439639": 60,
    "439640": 59,
    "439641": 59,
    "439642": 59,
    "439643": 58,
    "439644": 58,
    "439645": 58,
    "439646": 57,
    "439647": 57,
    "439648": 57,
    "439649": 56,
    "439650": 56,
    "439651": 56,
    "439652": 55,
    "439653": 55,
    "439654": 55,
    "439655": 54,
    "439656": 54,
    "439657": 54,
    "439658": 53,
    "439659": 53,
    "439660": 53,
    "439661": 52,
    "439662": 52,
    "439663": 52,
    "439664": 51,
    "439665": 51,
    "439666": 51,
    "439667": 50,
    "439668": 50,
    "439669": 50,
    "439670": 49,
    "439671": 49,
    "439672": 49,
    "439673": 48,
    "439674": 48,
    "439675": 48,
    "439676": 47,
    "439677": 47,
    "439678": 47,
    "439679": 46,
    "439680": 46,
    "439681": 46,
    "439682": 45,
    "439683": 45,
    "439684": 45,
    "439685": 44,
    "439686": 44,
    "439687": 44,
    "439688": 43,
    "439689": 43,
    "439690": 43,
    "439691": 42,
    "439840": 42,
    "439841": 42,
    "439842": 41,
    "439843": 41,
    "439844": 41,
    "439845": 40,
    "439846": 40,
    "439847": 40,
    "439848": 39,
    "439849": 39,
    "439850": 39,
    "439851": 38,
    "439852": 38,
    "439853": 38,
    "439854": 37,
    "439855": 37,
    "439856": 37,
    "439857": 36,
    "439858": 36,
    "439859": 36,
    "439860": 35,
    "439861": 35,
    "439862": 35,
    "439863": 34,
    "439864": 34,
    "439865": 34,
    "439866": 33,
    "439867": 33,
    "439868": 33,
    "439869": 32,
    "439870": 32,
    "439871": 32,
    "439872": 31,
    "439873": 31,
    "439874": 31,
    "439875": 30,
    "439876": 30,
    "439877": 30,
    "439878": 29,
    "439879": 29,
    "439880": 29,
    "439881": 28,
    "439882": 28,
    "439883": 28,
    "439884": 27,
    "439885": 27,
    "439886": 27,
    "439887": 26,
    "439888": 26,
    "439889": 26,
    "439890": 25,
    "439891": 25,
    "439892": 25,
    "439893": 24,
    "439894": 24,
    "439895": 24,
    "439896": 23,
    "439897": 23,
    "439898": 23,
    "439899": 22,
    "439900": 22,
    "439901": 22,
    "439902": 21,
    "439903": 21,
    "439904": 21,
    "439905": 20,
    "439906": 20,
    "439907": 20,
    "439908": 19,
    "439909": 19,
    "439910": 19,
    "439911": 18,
    "439912": 18,
    "439913": 18,
    "439914": 17,
    "439915": 17,
    "439916": 17,
    "439917": 16,
    "439918": 16,
    "439919": 16,
    "439920": 15,
    "439921": 15,
    "439922": 15,
    "439923": 14,
    "439924": 14,
    "439925": 14,
    "439926": 13,
    "439927": 13,
    "439928": 13,
    "439929": 12,
    "439930": 12,
    "439931": 12,
    "439932": 11,
    "439933": 11,
    "439934": 11,
    "439935": 10,
    "440063": 10,
    "440064": 10,
    "440065": 9,
    "440066": 9,
    "440067": 9,
    "440068": 8,
    "440069": 8,
    "440070": 8,
    "440071": 7,
    "440072": 7,
    "440073": 7,
    "440074": 6,
    "440075": 6,
    "440076": 6,
    "440077": 5,
    "440078": 5,
    "440079": 5,
    "440080": 4,
    "440081": 4,
    "440082": 4,
    "440083": 3,
    "440084": 3,
    "440085": 3,
    "440086": 2,
    "440087": 2,
    "440088": 2,
    "440089": 1,
    "440090": 1,
    "440091": 1,
    "440092": 0,
    "440093": 0,
    "440094": 0,
  },
  "p2g": {
    "0": 440094,
    "1": 440091,
    "10": 440064,
    "100": 438661,
    "101": 438658,
    "102": 438655,
    "103": 438652,
    "104": 438649,
    "105": 438646,
    "106": 438643,
    "107": 438640,
    "108": 438319,
    "109": 438316,
    "11": 439934,
    "110": 438313,
    "111": 438310,
    "112": 438307,
    "113": 438304,
    "114": 438301,
    "115": 438298,
    "116": 438295,
    "117": 438292,
    "12": 439931,
    "13": 439928,
    "14": 439925,
    "15": 439922,
    "16": 439919,
    "17": 439916,
    "18": 439913,
    "19": 439910,
    "2": 440088,
    "20": 439907,
    "21": 439904,
    "22": 439901,
    "23": 439898,
    "24": 439895,
    "25": 439892,
    "26": 439889,
    "27": 439886,
    "28": 439883,
    "29": 439880,
    "3": 440085,
    "30": 439877,
    "31": 439874,
    "32": 439871,
    "33": 439868,
    "34": 439865,
    "35": 439862,
    "36": 439859,
    "37": 439856,
    "38": 439853,
    "39": 439850,
    "4": 440082,
    "40": 439847,
    "41": 439844,
    "42": 439841,
    "43": 439690,
    "44": 439687,
    "45": 439684,
    "46": 439681,
    "47": 439678,
    "48": 439675,
    "49": 439672,
    "5": 440079,
    "50": 439669,
    "51": 439666,
    "52": 439663,
    "53": 439660,
    "54": 439657,
    "55": 439654,
    "56": 439651,
    "57": 439648,
    "58": 439645,
    "59": 439642,
    "6": 440076,
    "60": 439639,
    "61": 439636,
    "62": 439633,
    "63": 439630,
    "64": 439627,
    "65": 439624,
    "66": 439621,
    "67": 439618,
    "68": 439615,
    "69": 439612,
    "7": 440073,
    "70": 439609,
    "71": 439606,
    "72": 439603,
    "73": 439600,
    "74": 439597,
    "75": 439594,
    "76": 439591,
    "77": 439588,
    "78": 439585,
    "79": 439582,
    "8": 440070,
    "80": 439437,
    "81": 439434,
    "82": 439431,
    "83": 439428,
    "84": 439425,
    "85": 439422,
    "86": 439419,
    "87": 439416,
    "88": 439413,
    "89": 439410,
    "9": 440067,
    "90": 439407,
    "91": 439404,
    "92": 439401,
    "93": 439398,
    "94": 439395,
    "95": 439392,
    "96": 439389,
    "97": 438670,
    "98": 438667,
    "99": 438664,
  },
  "refName": "NC_036159.2",
  "strand": -1,
}
`;
